rfcs/syntax-static-imports.md

```diff
@@ -0,0 +1,281 @@
+!import Package:WaitForChild("Libraries"):WaitForChild("Fusion")
+```
+
+### Syntax: Basic form
+
+The most basic form gives only the expression which resolves statically to the module path:
+
+```Lua
+!import "foo/bar/baz"
+```
+
+The returned value from the module is placed in a local variable, adopting the module's name. Types are placed in a namespace of the same name. This is how `require()` works today.
+
+The above snippet can be seen as equivalent to:
+
+```Lua
+local baz = require("foo/bar/baz")
+```
+
+### Syntax: Renaming
+
+To use a different name for the local variable/type namespace, an equals sign is added at the start of the statement:
+
+```Lua
+!import not_baz = "foo/bar/baz"
+```
+
+Beyond user convenience, this allows code to deal with modules named identically, as can be done today with `require()`.
+
+The above snippet can be seen as equivalent to:
+
+```Lua
+local not_baz = require("foo/bar/baz")
+```
+
+### Syntax: Type prefix
+
+To import only the types from a module, and skip evaluating the module at runtime, the `type` keyword can be added.
+
+```Lua
+!import type "foo/bar/baz"
+!import not_baz = type "foo/bar/baz"
+```
+
+This allows type information to be fetched from a module without introducing cyclic dependencies at runtime.
+
+There is currently no exactly equivalent code snippet for this. The closest is ['type smuggling' as presented by Anaminus](https://twitter.com/Anaminus/status/1585287008661938180), which does not work with exported types:
+
+```Lua
+type baz = typeof(require("foo/bar/baz"))
+```
+
+### Syntax: Local prefix
+
+To destructure an import and insert its contents directly into the current namespace, the `local` keyword can be added (before `type` if present):
+
+```Lua
+!import local "foo/bar/baz"
+!import local type "foo/bar/baz"
+```
+
+This is especially useful in the case of DSL-like libraries, or any libraries that wish to include a prelude of commonly used members. It is acknowledged that this can lead to namespace pollution, but this is something the developer is in control of at all times, and explicitly opts into.
+
+Unless only types are being imported, the module must return a table. All statically resolvable members of the table, which have string keys and are valid identifiers, are turned into local variables in the current namespace.
+
+While there is no syntax ambiguity, the `local` prefix is not sensible with renaming, because it does not make sense to rename a namespace that will not be created. This case should likely warn, but is not necessarily a failure case.
+
+```Lua
+!import not_baz = local "foo/bar/baz" -- why?
+```
+
+There is no equivalent code snippet, though similar behaviour without type importing can be achieved with unidiomatic use of `getfenv()`:
+
+```Lua
+local _temp = require("foo/bar/baz")
+for ident, value in _temp do
+    getfenv()[ident] = value
+end
+_temp = nil
+```
+
+### Syntax: Member list
```

Comment:
This should definitely just be destructuring, which is applicable in far more contexts.

Reply:
My general stance is that:

- I really like destructuring
- I think destructuring can solve many of the same problems this RFC motivates
- I am concerned that it might not be feasible to implement general destructuring because the RFC has thrown up some pretty deep issues.
- As a result, I propose this as a way of specifically cleaning up the use cases around imports and avoiding the problems that RFC presents.

I would agree that destructuring would work. If it ever gets implemented. Which is not a given.

Reply:
I have elaborated on my thoughts on destructuring below.